features/features.go

```diff
@@ -0,0 +1,75 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package features
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+// SnapdFeature is a named feature that may be on or off.
+type SnapdFeature int
+
+const (
+	// Layouts controls availability of snap layouts.
+	Layouts SnapdFeature = iota
+	// ParallelInstances controls availability installing a snap multiple times.
+	ParallelInstances
+	// Hotplug controls availability of dynamically creating slots based on system hardware.
+	Hotplug
+	// SnapdSnap controls possibility of installing the snapd snap.
+	SnapdSnap
+	// PerUserMountNamespace controls the persistence of per-user mount namespaces.
+	PerUserMountNamespace
+	// PerformanceMeasurements controls the collection of performance data.
+	PerformanceMeasurements
+)
+
+// String returns the name of a snapd feature.
+func (f SnapdFeature) String() string {
+	// The constants here must be synchronized with cmd/libsnap-confine-private/feature.c
```

Comment:
And they also need to be synced with overlord/configstate/configcore/experimental.go  - this makes me a bit uneasy. I wonder if there is a way to sync configcore and this closer. Right now we when adding a experimental flag we need to add things in three places AFAICT, I wonder if there is a way to share the GO bits more (and maybe even the C bits but that is of course way harder).

Reply:
As I said in the other PR I will follow up with using those constants there. We will only need to do two places (C and Go) which is IMO good enough for now.

Reply:
we are growing a bit too many small packages I fear.  If we have this the code in configcore/experimental.go should probably become dynamic and use listing here that should flag a feature as experimental and whether it's turned to default.